src/main/java/com/conveyal/r5/profile/FrequencyRandomOffsets.java

```diff
@@ -179,10 +224,27 @@ public void randomize () {
                     }
                 }
             }
-
-            if (remainingAfterPreviousRound == remaining && remaining > 0) {
+            if (remainingAfterPreviousPass == remaining && remaining > 0) {
                 throw new IllegalArgumentException("Cannot solve phasing, you may have a circular reference!");
             }
+            // Copy results of randomization to a Map keyed on TripSchedules (instead of TripPattern index ints). This
+            // allows looking up offsets in a context where we only have a filtered list of running frequency trips and
+            // don't know the original unfiltered index of the trip within the pattern. Ideally we'd just build the
+            // original map keyed on TripSchedules (or hypothetical FreqEntries) but that reqires a lot of refactoring.
+            offsetsForTripSchedule.clear();
+            for (TIntObjectIterator<int[][]> it = offsets.iterator(); it.hasNext(); ) {
+                it.advance();
+                TripPattern tripPattern = data.tripPatterns.get(it.key());
+                int[][] offsetsForTrip = it.value();
+                for (int ts = 0; ts < tripPattern.tripSchedules.size(); ts++) {
+                    TripSchedule tripSchedule = tripPattern.tripSchedules.get(ts);
+                    // On patterns with mixed scheduled and frequency trips, scheduled trip slots will be null.
+                    // Maps can store null values, but there's no point in storing them. We only store non-null arrays.
```

Comment:
Note to self: reads from this map occur only in `getOffsetSeconds (TripSchedule tripSchedule, int freqEntryIndex)`, which is only called when iterating over trips that have frequency entries. So indeed, there is no point in storing a `null` value that will never be read.